### PR TITLE
doc/website/blog: Generate a fully-working Go CRUD HTTP API with Ent

### PR DESCRIPTION
**Pull request by masseelch**
Have a final look please. :heart:

### CODE REVIEW

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+Today, we would like to present a new Ent extension named `elk` that can automatically generate fully-working, RESTful
+API endpoints from your Ent schemas. `elk` strives to automate all of the tedious work of setting up the basic CRUD
+endpoints for every entity you add to your graph, including logging, validation of the request body, eager loading
+relations and serializing, all while leaving reflection out of sight and maintaining type-safety.
+
+Let’s get started!
+
+### Getting Started
+
+The final version of the code below can be found on GitHub.
```
Comment (a8m) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:31]:
```suggestion
The final version of the code below can be found on [GitHub](https://github.com/masseelch/elk-example).
```

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+    if err != nil {
+        l.Error("serialization error", zap.Int("id", id), zap.Error(err))
+        render.InternalServerError(w, r, nil)
+        return
+    }
+    l.Info("pet rendered", zap.Int("id", id))
+    render.OK(w, r, d)
+}
+```
+<!--// @formatter:on-->
```
Comment (a8m) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:229]: ??
```suggestion
```

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+	"elk-example/ent"
+	elk "elk-example/ent/http"
+	"fmt"
+	"log"
+	"net/http"
+
+	"github.com/go-chi/chi/v5"
+	"github.com/go-playground/validator/v10"
+	_ "github.com/mattn/go-sqlite3"
+	"go.uber.org/zap"
```
Comment (a8m) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:248]:
```suggestion
	"context"
	"fmt"
	"log"
	"net/http"
	
	"elk-example/ent"
	elk "elk-example/ent/http"

	"github.com/go-chi/chi/v5"
	"github.com/go-playground/validator/v10"
	_ "github.com/mattn/go-sqlite3"
	"go.uber.org/zap"
```

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+func (Pet) Edges() []ent.Edge {
+    return []ent.Edge{
+        edge.From("owner", User.Type).
+            Ref("pets").
+            Unique(),
+    }
+}
+
+```
+<!--// @formatter:on-->
```
Comment (a8m) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:333]:
```suggestion
```

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+
+### Relations
+
+To illustrate more of `elk`s features, let’s extend our graph. Edit `ent/schema/user.go` and `ent/schema/pet.go`:
+
+`ent/schema/pet.go`:
+
+<!--// @formatter:off-->
+```go
+// Edges of the Pet.
```
Comment (a8m) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:323]: Why the `<!--// @formatter:off-->` comment is needed?

```suggestion
```go title="ent/schema/pet.go"
// Edges of the Pet.
```

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+
+```shell
+go generate ./...
+```
+
+Observe that in addition to the files Ent would normally generate, another directory named `ent/http` was created. These
+files were generated by the `elk` extension and they contain the code for the generated HTTP handlers. For example, here
+is some of the generated code for a read-operation on the Pet entity:
+
+<!--// @formatter:off-->
```
Comment (a8m) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:158]:
```suggestion
```

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+            Unique(),
+    }
+}
+
+```
+<!--// @formatter:on-->
+
+`ent/schema/user.go`:
+
+```go
```
Comment (a8m) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:337]:
```suggestion
```go title="ent/schema/user.go"
```

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+
+Rerun the code generator:
+
+```shell
+go generate ./...
+```
+
+Do not forget to register the `UserHandler` on our router. Just add the following lines to `main.go`:
+
+<!--// @formatter:off-->
```
Comment (a8m) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:379]:
```suggestion
```

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 9]
```diff
@@ -0,0 +1,550 @@
...
+    })
++    // Create the user handler.
++    r.Route("/users", func(r chi.Router) {
++        elk.NewUserHandler(c, l, v).Mount(r, elk.UserRoutes)
++    })
+    // Start listen to incoming requests.
+    fmt.Println("Server running")
+[...]
+```
+<!--// @formatter:on-->
```
Comment (a8m) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:393]:
```suggestion
```

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 10]
```diff
@@ -0,0 +1,550 @@
...
+  "name": "Elk"
+}
+```
+
+From the output we can see that the user has been created, but the edges are empty. `elk` does not include edges in its
+output by default. You can configure `elk` to render edges using a feature called “serialization groups”. Annotate your
+schemas with the `elk.SchemaAnnotation`
+and `elk.Annotation` structs. Edit `ent/schema/user.go` and add those:
+
+<!--// @formatter:off-->
```
Comment (a8m) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:417]:
```suggestion
```

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 11]
```diff
@@ -0,0 +1,550 @@
...
+    }
+}
+
+// Annotations of the User.
+func (User) Annotations() []schema.Annotation {
+    return []schema.Annotation{elk.ReadGroups("user")}
+}
+
+```
+<!--// @formatter:on-->
```
Comment (a8m) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:433]:
```suggestion
```

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 12]
```diff
@@ -0,0 +1,550 @@
...
+
+Our current schemas allow to set a negative age for pets or users and we can create pets without an owner (as we did
+with Kuro). Ent has built-in support for basic [validation](https://entgo.io/docs/schema-fields#validators). In some
+cases you may want to validate requests made against your API before passing their payload to Ent. `elk`
+uses [this](https://github.com/go-playground/validator) package to define validation rules and validate data. We can
+create separate validation rules for Create and Update operations using `elk.Annotation`. In our example, let’s assume
+that we want our Pet schema to only allow ages greater than zero and to disallow creating a pet without an owner.
+Edit `ent/schema/pet.go`:
+
+<!--// @formatter:off-->
```
Comment (a8m) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:475]:
```suggestion
```

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 13]
```diff
@@ -0,0 +1,550 @@
...
+    return []ent.Edge{
+        edge.From("owner", User.Type).
+            Ref("pets").
+            Unique().
+            Required().
+            Annotations(elk.Validation("required")),
+    }
+}
+```
+<!--// @formatter:on-->
```
Comment (a8m) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:498]:
```suggestion
```

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 14]
```diff
@@ -0,0 +1,550 @@
...
+
+<!--// @formatter:off-->
+```go
+// Fields of the Pet.
+func (Pet) Fields() []ent.Field {
+    return []ent.Field{
+        field.String("name"),
+        field.Int("age").
+            Positive().
+            Annotations(elk.CreateValidation("required,gt=0"),elk.UpdateValidation("gt=0")),
```
Comment (a8m) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:483]:
```suggestion
            Annotations(
                elk.CreateValidation("required,gt=0"),
                elk.UpdateValidation("gt=0"),
            ),
```

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 15]
```diff
@@ -0,0 +1,550 @@
...
+request validation is especially useful if you'd wanted to do cross-field-validation.
+
+### Upcoming Features
+
+We hope you agree that `elk` has some useful features already, but there are still many exciting things to come. The
+next version of `elk` will include::
+
+- Fully working flutter frontend to administrate your nodes
+- Integration of Ent’s validation in the current request validator
+- More transport formats (currently only json)
```
Comment (a8m) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:535]:
```suggestion
- More transport formats (currently only JSON)
```

**Review by a8m — APPROVED** [review]

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 16]
```diff
@@ -0,0 +1,550 @@
...
+
+We added two fields to our `Pet` entity: `name` and `age`. The `ent.Schema` just defines the fields of our entity. To
+generate runnable code from our schema, run:
+
+```shell
+go generate ./...
+```
+
+Observe that in addition to the files Ent would normally generate, another directory named `ent/http` was created. These
+files were generated by the `elk` extension and they contain the code for the generated HTTP handlers. For example, here
```
Comment (rotemtam) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:155]:
```suggestion
files were generated by the `elk` extension and contain the code for the generated HTTP handlers. For example, here
```

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 17]
```diff
@@ -0,0 +1,550 @@
...
+{
+  "age": 30,
+  "edges": {},
+  "id": 1,
+  "name": "Elk"
+}
+```
+
+From the output we can see that the user has been created, but the edges are empty. `elk` does not include edges in its
+output by default. You can configure `elk` to render edges using a feature called “serialization groups”. Annotate your
```
Comment (rotemtam) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:413]:
```suggestion
output by default. You can configure `elk` to render edges using a feature called "serialization groups". Annotate your
```

File: doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md [hunk 18]
```diff
@@ -0,0 +1,550 @@
...
+some repetitive tasks that go into building RESTful APIs and focus on more meaningful work.
+
+`elk` is in an early stage of development, we welcome any suggestion or feedback and if you are willing to help we'd be
+very glad. The [GitHub Issues](https://github.com/masseelch/elk/issues) is a wonderful place for you to reach out for
+help, feedback, suggestions and contribution.
+
+### About the Developer
+
+MasseElch is a software engineer in the windy, flat north of Germany. When not hiking with his dog Kuro (who has his own
+instagram channel :scream:) or playing hide-and-seek with his son, he drinks coffee and codes.
```
Comment (rotemtam) [inline doc/website/blog/2021-09-29-generate-a-fully-working-go-crud-http-api-with-ent.md:550]:
```suggestion
#### About the Author

_MasseElch is a software engineer from the windy, flat north of Germany. When not hiking with his dog Kuro (who has his own Instagram channel :scream:) or playing hide-and-seek with his son, he drinks coffee and enjoys coding._
```

**Pull request merged**